src/test/osd/RadosModel.h

```diff
@@ -1059,15 +1068,48 @@ class ReadOp : public TestOp {
     }
     op.getxattrs(&xattrs, 0);
     assert(!context->io_ctx.aio_operate(context->prefix+oid, completion, &op, 0));
+    waiting_on++;
+ 
+    // send 2 pipelined reads on the same object/snap. This can help testing
+    // OSD's read behavior in some scenarios
+    for (uint32_t i = 0; i < 2; ++i) {
+      librados::ObjectReadOperation pipeline_op;
+
+      pipeline_op.read(0,
+                       !old_value.has_contents() ? 0 :
+                       old_value.most_recent_gen()->get_length(old_value.most_recent()),
+                       &pipeline_results[i],
+                       &pipeline_retvals[i]);
+      pipeline_comps[i] = context->rados.aio_create_completion((void *) this, &read_callback, 0);
+      assert(!context->io_ctx.aio_operate(context->prefix+oid, pipeline_comps[i], &pipeline_op, 0));
+      waiting_on++;
+    }
+
     if (snap >= 0) {
       context->io_ctx.snap_set_read(0);
     }
+    context->state_lock.Unlock();
   }
 
   void _finish(CallbackInfo *info)
   {
-    context->state_lock.Lock();
+    Mutex::Locker l(context->state_lock);
     assert(!done);
+    assert(waiting_on > 0);
+    if (--waiting_on) {
+      return;
+    }
+
+    for (vector<librados::AioCompletion *>::iterator i = pipeline_comps.begin();
+         i != pipeline_comps.end(); ++i) {
+      assert((*i)->is_complete());
+      if (int err = (*i)->get_return_value()) {
```

Comment:
@wonzhq , sorry i didn't have a good enough understanding of the RadosModel yesterday, when i am looking at your improvement, i see following we can probably improve in following ways:
- is it possible that the old_value is deleted? so we might want to take the case of `(err == -ENOENT && old_value.deleted())` as expected.
- probably we can `assert(0)` at seeing an unexpected return value? i see we are doing this for the `completion->get_return_value()`.
- do you want to check the returned bufferlist in addition to the "return value" and count the # of mismatched returned buffer using `context->errors`? IMO, it's also expected to have the same buffer list as the result of the pipelined read op.
- probably we can consolidate the check for `pipeline_results` with `result`, if we do not plan to distinguish the number of errors of pipelined (read) op from the ordinary op? and also the check and life management for the completion objects. because they share the same test conditions. this is more of a refactory which merge the `completion`, `result`, `retval` and `pipeline_comps`, `pipeline_results`, `pipeline_retval` into `completions`, `results`,  `retvals`, so if you are good with this idea, this refactor deserves another commit in this pull request.

thanks. =)